caas/kubernetes/provider/rbac.go

```diff
@@ -0,0 +1,462 @@
+}
+
+func (k *kubernetesClient) deleteServiceAccountsRolesBindings(appName string) error {
+	listOps := v1.ListOptions{
+		LabelSelector: applicationSelector(appName),
+	}
+	rbList, err := k.client().RbacV1().RoleBindings(k.namespace).List(listOps)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	for _, s := range rbList.Items {
+		if err := k.deleteRoleBinding(s.GetName()); err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	crbList, err := k.client().RbacV1().ClusterRoleBindings().List(listOps)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	for _, s := range crbList.Items {
+		if err := k.deleteClusterRoleBinding(s.GetName()); err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	rList, err := k.client().RbacV1().Roles(k.namespace).List(listOps)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	for _, s := range rList.Items {
+		if err := k.deleteRole(s.GetName()); err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	crList, err := k.client().RbacV1().ClusterRoles().List(listOps)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	for _, s := range crList.Items {
+		if err := k.deleteClusterRole(s.GetName()); err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	saList, err := k.client().CoreV1().ServiceAccounts(k.namespace).List(listOps)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	for _, s := range saList.Items {
+		if err := k.deleteServiceAccount(s.GetName()); err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+func (k *kubernetesClient) createRole(role *rbacv1.Role) (*rbacv1.Role, error) {
+	out, err := k.client().RbacV1().Roles(k.namespace).Create(role)
+	if k8serrors.IsAlreadyExists(err) {
+		return nil, errors.AlreadyExistsf("role %q", role.GetName())
+	}
+	return out, errors.Trace(err)
+}
+
+func (k *kubernetesClient) updateRole(role *rbacv1.Role) (*rbacv1.Role, error) {
+	out, err := k.client().RbacV1().Roles(k.namespace).Update(role)
+	if k8serrors.IsNotFound(err) {
+		return nil, errors.NotFoundf("role %q", role.GetName())
+	}
+	return out, errors.Trace(err)
+}
+
+func (k *kubernetesClient) ensureRole(role *rbacv1.Role) (*rbacv1.Role, error) {
+	out, err := k.updateRole(role)
+	if errors.IsNotFound(err) {
+		out, err = k.createRole(role)
+	}
+	return out, errors.Trace(err)
+}
+
+func (k *kubernetesClient) getRole(name string) (*rbacv1.Role, error) {
+	out, err := k.client().RbacV1().Roles(k.namespace).Get(name, v1.GetOptions{IncludeUninitialized: true})
+	if k8serrors.IsNotFound(err) {
+		return nil, errors.NotFoundf("role %q", name)
+	}
+	return out, errors.Trace(err)
+}
+
+func (k *kubernetesClient) deleteRole(name string) error {
+	err := k.client().RbacV1().Roles(k.namespace).Delete(name, &v1.DeleteOptions{
+		PropagationPolicy: &defaultPropagationPolicy,
+	})
+	if k8serrors.IsNotFound(err) {
+		return nil
+	}
+	return errors.Trace(err)
+}
+
+func (k *kubernetesClient) deleteRoles(appName string) error {
+	roleList, err := k.client().RbacV1().Roles(k.namespace).List(v1.ListOptions{
```

Comment:
I'm not a huge fan of using the result-set from a List call with k8s selectors being used to delete or update. In this case it's probably fine, but if juju already knows what it should delete, we should use our source of truth.

My concern is that we are relying on kubernetes (and our selectors) to choose the correct objects to delete.

Reply:
Juju labels the resources when it creates the resources and selects the resources by labels to update or delete. So the label for caas, tag for iaas. 
I m not sure if I understand ur concerts.

Reply:
as discussed, we already ensure this by labelling the resources.